include/util/hex_float.h

```diff
@@ -213,12 +270,310 @@ class HexFloat {
+  const uint_type getNormalizedSignificand() const {
+    int_type unbiased_exponent = getUnbiasedNormalizedExponent();
+    uint_type significand = getSignificandBits();
+    for (int_type i = unbiased_exponent;
+         i <= -static_cast<int_type>(exponent_bias); ++i) {
+      significand <<= 1;
+    }
+    significand &= fraction_encode_mask;
+    return significand;
+  }
+
+  // Returns true if this number represents a negative value.
+  bool isNegative() const { return (getBits() & sign_mask) != 0; }
+
+  // Note this assumes EVERY significand is normalized, and has an implicit
+  // leading one. This means that the only way that this method will set 0,
+  // is if you set a number so denormalized that it underflows.
+  // The significand is also expected to be in the
+  // lowest-most num_fraction_bits of the uint_type.
+  // The exponent is expected to be unbiased, meaning an exponent of
+  // 0 actually means 0.
+  // If underflow_round_up is set, then on underflow, if a number is non-0
+  // and would underflow, we round up to the smallest denorm.
+  void setFromSignUnbiasedExponentAndNormalizedSignificand(
+      bool negative, int_type exponent, uint_type significand,
+      bool round_denorm_up) {
+    const int_type negative_bias = -static_cast<int_type>(exponent_bias);
+    bool significand_is_zero = significand == 0;
+
+    if (exponent <= negative_bias) {
+      // If this was dernom, then we have to shift the bit on, meaning
+      // the significand is not zero.
+      significand_is_zero = false;
+      significand |= first_exponent_bit;
+      significand >>= 1;
+    }
+
+    while (exponent < negative_bias) {
+      significand >>= 1;
+      ++exponent;
+    }
+
+    if (exponent == negative_bias) {
+      if (significand == 0 && !significand_is_zero && round_denorm_up) {
+        significand = 0x1;
+      }
+    }
+
+    uint_type value = 0;
+    if (negative) {
+      value |= sign_mask;
+    }
+    exponent += exponent_bias;
+    assert(exponent >= 0);
+
+    // put it all together
+    exponent = (exponent << exponent_left_shift) & exponent_mask;
+    significand &= fraction_encode_mask;
+    value |= exponent | significand;
+    value_ = BitwiseCast<T>(value);
+  }
+
+  // Increments the significand of this number by the given amount.
+  // If this would spill the significand into the implicit bit,
+  // carry is set to true and the significand is shifted to fit into
+  // the correct location.
+  static uint_type incrementSignificand(uint_type significand,
+                                         uint_type to_increment, bool* carry) {
+    significand += to_increment;
+    if (significand & first_exponent_bit) {
+      *carry = true;
+      // The implicit 1-bit will have carried, so we should zero-out the
+      // top bit and shift back.
+      significand &= ~first_exponent_bit;
+      significand >>= 1;
+    }
+    return significand;
+  }
+
+  // Returns the significand, rounded to fit in a significand of
```

Comment:
Blurb doesn't quite line up with the method signature.
The num_bits is the num_bits for the other_T ?